src/backend/renderer/pixman/mod.rs

```diff
@@ -0,0 +1,1183 @@
+            #[cfg(target_endian = "little")]
+            DrmFourcc::Abgr2101010,
+        ]
+    };
+}
+
+#[derive(Debug)]
+enum PixmanTarget {
+    Image { dmabuf: Dmabuf, image: PixmanImage },
+    RenderBuffer(PixmanRenderBuffer),
+}
+
+impl PixmanTarget {
+    fn image(&self) -> &pixman::Image<'static, 'static> {
+        match self {
+            PixmanTarget::Image { image, .. } => &image.0.image,
+            PixmanTarget::RenderBuffer(render_buffer) => &render_buffer.0,
+        }
+    }
+}
+
+/// Offscreen render buffer
+#[derive(Debug)]
+pub struct PixmanRenderBuffer(pixman::Image<'static, 'static>);
+
+impl From<pixman::Image<'static, 'static>> for PixmanRenderBuffer {
+    fn from(value: pixman::Image<'static, 'static>) -> Self {
+        Self(value)
+    }
+}
+
+#[derive(Debug)]
+struct PixmanDmabufMapping {
+    dmabuf: WeakDmabuf,
+    _mapping: DmabufMapping,
+}
+
+#[derive(Debug)]
+struct PixmanImageInner {
+    #[cfg(feature = "wayland_frontend")]
+    buffer: Option<Weak<wl_buffer::WlBuffer>>,
+    dmabuf: Option<PixmanDmabufMapping>,
+    image: Image<'static, 'static>,
+    _flipped: bool, /* TODO: What about flipped textures? */
+}
+
+#[derive(Debug, Clone)]
+struct PixmanImage(Rc<PixmanImageInner>);
+
+impl PixmanImage {
+    #[profiling::function]
+    fn accessor<'l>(&'l self) -> Result<TextureAccessor<'l>, PixmanError> {
+        #[cfg(feature = "wayland_frontend")]
+        let buffer = if let Some(buffer) = self.0.buffer.as_ref() {
+            Some(buffer.upgrade().map_err(|_| PixmanError::BufferDestroyed)?)
+        } else {
+            None
+        };
+
+        let guard = if let Some(mapping) = self.0.dmabuf.as_ref() {
+            let dmabuf = mapping.dmabuf.upgrade().ok_or(PixmanError::BufferDestroyed)?;
+            Some(DmabufReadGuard::new(dmabuf)?)
+        } else {
+            None
+        };
+
+        Ok(TextureAccessor {
+            #[cfg(feature = "wayland_frontend")]
+            buffer,
+            image: &self.0.image,
+            _guard: guard,
+        })
+    }
+}
+
+/// A handle to a pixman texture
+#[derive(Debug, Clone)]
+pub struct PixmanTexture(PixmanImage);
+
+impl From<pixman::Image<'static, 'static>> for PixmanTexture {
+    fn from(image: pixman::Image<'static, 'static>) -> Self {
+        Self(PixmanImage(Rc::new(PixmanImageInner {
+            #[cfg(feature = "wayland_frontend")]
+            buffer: None,
+            dmabuf: None,
+            _flipped: false,
+            image,
+        })))
+    }
+}
+
+struct DmabufReadGuard {
+    dmabuf: Dmabuf,
+}
+
+impl DmabufReadGuard {
+    #[profiling::function]
+    pub fn new(dmabuf: Dmabuf) -> std::io::Result<Self> {
+        dmabuf.sync(DmaBufSyncFlags::START | DmaBufSyncFlags::READ)?;
+        Ok(Self { dmabuf })
+    }
+}
+
+impl Drop for DmabufReadGuard {
+    #[profiling::function]
+    fn drop(&mut self) {
+        if let Err(err) = self.dmabuf.sync(DmaBufSyncFlags::END | DmaBufSyncFlags::READ) {
```

Comment:
Should this have `WRITE` also in case something happened after a SYNC_WRITE causes a drop?

Reply:
The `DmabufReadGuard` is only used for imported dmabufs, so only ever reads and never issues a WRITE